### PR TITLE
Bug 1109684 - Let other developers directly open pages in Firefox for iOS (Part 2)

### PR DESCRIPTION
**Pull request by bkmunar**
Implementing callback functionality to 3rd party apps

### CODE REVIEW

**Comment by bkmunar** [comment]
wip

File: Client/Application/AppDelegate.swift [hunk 1]
```diff
@@ -85,6 +86,17 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         return true
     }
 
+    func backToApp() {
+
```
Comment (rnewman) [inline Client/Application/AppDelegate.swift]: I think you want a sanity guard here to make sure this runs on the main thread.

File: Client/Application/AppDelegate.swift [hunk 2]
```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
                     }
                 }
             }
+        } else if let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false) where components.scheme == "firefox" && components.host == "x-callback-url" {
```
Comment (rnewman) [inline Client/Application/AppDelegate.swift]: Let's clean this up, not cargo-cult it.

```
let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false)
if components.scheme != "firefox" {
    return false
}
switch components.host {
case "firefox":
…
case "x-…"
}
```

You get the idea. Too much nesting, too confused.

File: Client/Application/AppDelegate.swift [hunk 3]
```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+            if let query = components.query, items = components.queryItems as? [NSURLQueryItem] {
+                if items[0].name == "x-source" && items[0].value != nil && items[1].name == "url" && items[1].value != nil {
```
Comment (rnewman) [inline Client/Application/AppDelegate.swift]: There has got to be a better way to do this.

File: Client/Application/AppDelegate.swift [hunk 4]
```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+                    if let newURL = NSURL(string: items[1].value!) {
+                        let tab = self.tabManager.addTab(request: NSURLRequest(URL: newURL))
+                        self.tabManager.selectTab(tab)
```
Comment (rnewman) [inline Client/Application/AppDelegate.swift]: Does this update the tab counter? My understanding is that it does not.

File: Client/Application/AppDelegate.swift [hunk 5]
```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+                        if let backToAppURL = NSURL(string: items[0].value!) {
+                            if let appName = items[2].value where items[2].name == "x-source-name" {
```
Comment (rnewman) [inline Client/Application/AppDelegate.swift]: Don't pyramid. Swift lets you do this:

```
if let foo = bar,
       baz = noo where noo.foo {
}
```

File: Client/Application/AppDelegate.swift [hunk 6]
```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+                                if UIApplication.sharedApplication().canOpenURL(backToAppURL) {
```
Comment (rnewman) [inline Client/Application/AppDelegate.swift]: Invert this check:

```
if !UIApplication.s… {
    // Can't open this app, so just open the page without the app-back button.
    return true
}
```

File: Client/Application/AppDelegate.swift [hunk 7]
```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+
+                                    //button code will change when we use the button with image asset
+                                    var button = UIButton()
+                                    button.setTitle("〈 \(appName)", forState: UIControlState.Normal)
+                                    var fontSize: CGFloat = 14
+                                    button.titleLabel!.font = UIFont(name: "HelveticaNeueInterface", size: fontSize)
+                                    button.setTitleColor(UIColor.grayColor(), forState: UIControlState.Normal)
+                                    button.titleLabel!.lineBreakMode = NSLineBreakMode.ByTruncatingTail
+                                    button.addTarget(self, action: "backToApp", forControlEvents: UIControlEvents.TouchUpInside)
```
Comment (rnewman) [inline Client/Application/AppDelegate.swift]: Break these out into a method: `makeBackButtonForApp:` or something.

File: Client/Application/AppDelegate.swift [hunk 8]
```diff
@@ -96,6 +108,36 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+
+                                    tab.callbackToAppButton = button
+                                    tab.callbackURL = backToAppURL
+                                    self.browserViewController.urlBar.addSubview(button)
+                                    self.browserViewController.urlBar.updateConstraints()
```
Comment (rnewman) [inline Client/Application/AppDelegate.swift]: This is kinda wrong. You're directly manipulating the `tabManager` above, and you're manipulating the controller here.

What we really want is either:

* Make the tab, give it the attribute, and then switch to it — the right UI elements will appear as a result of switching
* Ask the BVC to create the tab with the right attribute and display.

You're splitting the difference between the two, which means there's code in this AppDelegate that constructs a button, but you also need to understand and touch BVC.

File: Client/Frontend/Browser/Browser.swift [hunk 9]
```diff
@@ -180,6 +182,13 @@ class Browser: NSObject {
         return nil
     }
 
+    var displayCallbackToAppButton: UIButton? {
+        if let callbackToAppButton = callbackToAppButton {
```
Comment (rnewman) [inline Client/Frontend/Browser/Browser.swift]: `self.callbackToAppButton` is equivalent to this entire method.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 10]
```diff
@@ -27,7 +27,7 @@ private struct BrowserViewControllerUX {
 
 class BrowserViewController: UIViewController {
 
-    private var urlBar: URLBarView!
+    var urlBar: URLBarView!
```
Comment (rnewman) [inline Client/Frontend/Browser/BrowserViewController.swift]: That you've had to shed the `private` nature of this is a sign you're doing something wrong. Don't break encapsulation.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 11]
```diff
@@ -742,6 +748,18 @@ extension BrowserViewController: URLBarDelegate {
         hideSearchController()
         updateInContentHomePanel(tabManager.selectedTab?.url)
     }
+
+    func urlBarUpdateCallbackToAppButtonBarState(urlBar: URLBarView) {
+        if let callbackURL = self.tabManager.selectedTab?.displayCallbackToAppButton {
```
Comment (rnewman) [inline Client/Frontend/Browser/BrowserViewController.swift]: Perhaps

```
if let callback = self.tabManager.selectedTab?.callbackToAppButton {
     callback.hidden = false
     self.urlBar.callbackToAppButton = callback
} else {
    if let existing = self.urlBar.callbackToAppButton {
        existing.hidden = true
    }
    self.urlBar.callbackToAppButton = nil
}
```

File: Client/Frontend/Browser/TabManager.swift [hunk 12]
```diff
@@ -181,6 +181,9 @@ class TabManager : NSObject {
         assert(NSThread.isMainThread())
         // If the removed tab was selected, find the new tab to select.
         if tab === selectedTab {
+            if let callbackToAppButton = tab.callbackToAppButton {
+                tab.callbackToAppButton = nil
```
Comment (rnewman) [inline Client/Frontend/Browser/TabManager.swift]: Just assign it, don't check.

File: Client/Application/AppDelegate.swift (hunk 4 above)
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: it does! just checked

File: Client/Frontend/Browser/TabManager.swift (hunk 12 above)
Comment (bkmunar) [inline Client/Frontend/Browser/TabManager.swift]: fixed

File: Client/Application/AppDelegate.swift (hunk 2 above)
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: fixed

File: Client/Application/AppDelegate.swift (hunk 7 above)
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: made the separate button class for it and included the function there

File: Client/Application/AppDelegate.swift (hunk 1 above)
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: included the function in the custom button class

File: Client/Application/AppDelegate.swift (hunk 5 above)
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: fixed

File: Client/Application/AppDelegate.swift (hunk 6 above)
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: this is checking to see if the app can be called back from firefox; I already open the tab before the backToAppURL check

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 10 above)
Comment (bkmunar) [inline Client/Frontend/Browser/BrowserViewController.swift]: fixed -- yeah this is why i needed feedback haha i wasn't sure how to follow the coding paradigm, but now it's really clear to me, thanks!

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 11 above)
Comment (bkmunar) [inline Client/Frontend/Browser/BrowserViewController.swift]: fixed

File: Client/Frontend/Browser/Browser.swift (hunk 9 above)
Comment (bkmunar) [inline Client/Frontend/Browser/Browser.swift]: took it out

File: Client/Application/AppDelegate.swift (hunk 8 above)
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: fixed -- yeah this is why i needed feedback haha i wasn't sure how to follow the coding paradigm, but now it's really clear to me, thanks!

**Comment by rnewman** [comment]
@ncalexan, @sleroux: could you take over reviewing this?

File: Client/Application/AppDelegate.swift [hunk 13]
```diff
@@ -5,12 +5,12 @@
 import Shared
 import Storage
 import AVFoundation
+import SnapKit
```
Comment (thebnich) [inline Client/Application/AppDelegate.swift]: AppDelegate shouldn't have any layout dependencies, so remove this.

File: Client/Application/AppDelegate.swift [hunk 14]
```diff
@@ -86,16 +86,39 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     }
 
     func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject?) -> Bool {
-        if let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false) where components.scheme == "firefox" && components.host == "open-url" {
-            if let query = components.query, item = components.queryItems?.first as? NSURLQueryItem {
-                if item.name == "url" && item.value != nil {
-                    if let newURL = NSURL(string: item.value!) {
-                        let tab = self.tabManager.addTab(request: NSURLRequest(URL: newURL))
-                        self.tabManager.selectTab(tab)
+        let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false)
```
Comment (thebnich) [inline Client/Application/AppDelegate.swift]: What is this `resolvingAgainstBaseURL: false` bit doing?

File: Client/Application/AppDelegate.swift [hunk 15]
```diff
@@ -86,16 +86,39 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+        if components!.scheme != "firefox" && components!.scheme != "firefox-x-callback" {
+            return false
+        }
+        switch components!.host! {
```
Comment (thebnich) [inline Client/Application/AppDelegate.swift]: Nit: Newline above here. It helps readability to add newlines after `return`s, and also before `if`/`while`/`switch` blocks (in this case, you have both).

**Comment by rnewman** [comment]
@bkmunar: could you clean up this PR, rebase it, and get some more feedback? It's best if you rebase rather than merging from master.

**Comment by fxios-buildasaur** [comment]
Result of integration 1
Integration took 9 minutes and 29 seconds.
All 0 tests passed, but please fix 18 warnings.

**Comment by thebnich** [comment]
Part 1 landed; part 2 no longer relevant AFAIK due to iOS 9 changes.

**Pull request closed without merging**